File: sqb-api-skills/sqb-pay/reference/pay_example.py
"""
收钱吧 B扫C 付款码支付示例 (Python)

核心流程：扫码 → 支付 → 轮询查询 → 确认结果
注意：所有交易都是真实的（无沙盒环境），测试后务必退款
"""

import hashlib
import json
import time
import uuid
from datetime import datetime

import requests

API_BASE = "https://vsi-api.shouqianba.com"


def md5_sign(body_str: str, key: str) -> str:
    """计算 MD5 签名"""
    content = body_str + key
    return hashlib.md5(content.encode("utf-8")).hexdigest()


def generate_client_sn(store_id: str = "STORE01") -> str:
    """生成唯一订单号"""
    timestamp = datetime.now().strftime("%Y%m%d%H%M%S%f")[:17]
    return f"{store_id}{timestamp}{uuid.uuid4().hex[:4]}"


class ShouqianbaPayClient:
    """收钱吧支付客户端"""

    def __init__(self, terminal_sn: str, terminal_key: str):
        self.terminal_sn = terminal_sn
        self.terminal_key = terminal_key
        self.session = requests.Session()
        self.session.headers.update({
            "Content-Type": "application/json; charset=utf-8",
        })

    def _request(self, endpoint: str, body: dict) -> dict:
        """发送签名请求"""
        body_str = json.dumps(body, ensure_ascii=False)
        sign = md5_sign(body_str, self.terminal_key)

        headers = {
            "Authorization": f"{self.terminal_sn} {sign}",
        }

        resp = self.session.post(
            f"{API_BASE}{endpoint}",
            data=body_str.encode("utf-8"),
            headers=headers,
            timeout=30,
        )
        return resp.json()

    def pay(
        self,
        client_sn: str,
        total_amount: str,
        dynamic_id: str,
        subject: str,
        operator: str,
        **kwargs,
    ) -> dict:
        """
        付款码支付

        Args:
            client_sn: 商户唯一订单号
            total_amount: 金额（单位：分）
            dynamic_id: 顾客付款码（扫码枪获取）
            subject: 交易简介（显示在顾客账单）
            operator: 操作员

        Returns:
            {"status": "SUCCESS/FAIL/PENDING", "data": ..., "message": ...}
        """
        body = {
            "terminal_sn": self.terminal_sn,
            "client_sn": client_sn,
            "total_amount": total_amount,
            "dynamic_id": dynamic_id,
            "subject": subject,
            "operator": operator,
            **kwargs,
        }

        result = self._request("/upay/v2/pay", body)
        return self._parse_pay_response(result, client_sn)

    def _parse_pay_response(self, result: dict, client_sn: str) -> dict:
        """三层状态判定"""
        # 第一层：通信层
        if result.get("result_code") != "200":
            return {
                "status": "FAIL",
                "data": None,
                "message": f"通信失败: {result.get('error_message')}",
            }

        biz = result.get("biz_response", {})
        biz_code = biz.get("result_code", "")

        # 第二层：业务层
        if biz_code == "PAY_FAIL":
            return {"status": "FAIL", "data": None, "message": "支付失败"}
        if biz_code in ("PAY_IN_PROGRESS", "PAY_FAIL_ERROR"):
            return {"status": "PENDING", "data": {"client_sn": client_sn}, "message": "支付处理中"}

        # 第三层：订单状态
        data = biz.get("data", {})
        order_status = data.get("order_status", "")

        if order_status == "PAID":
            return {"status": "SUCCESS", "data": data, "message": "支付成功"}
        elif order_status == "PAY_CANCELED":
            return {"status": "FAIL", "data": data, "message": "支付已撤销"}
        else:
            # CREATED, PAY_ERROR 等非最终状态
            return {"status": "PENDING", "data": data, "message": f"状态: {order_status}"}

    def query(self, client_sn: str = None, sn: str = None) -> dict:
        """查询订单"""
        body = {"terminal_sn": self.terminal_sn}
        if sn:
            body["sn"] = sn
        elif client_sn:
            body["client_sn"] = client_sn
        return self._request("/upay/v2/query", body)

    def pay_with_polling(
        self,
        client_sn: str,
        total_amount: str,
        dynamic_id: str,
        subject: str,
        operator: str,
        max_wait_seconds: int = 120,
        **kwargs,
    ) -> dict:
        """
        带轮询的完整支付流程

        Returns:
            {"status": "SUCCESS/FAIL/TIMEOUT", "data": ..., "message": ...}
        """
        # 1. 发起支付
        result = self.pay(client_sn, total_amount, dynamic_id, subject, operator, **kwargs)

        # 2. 最终状态直接返回
        if result["status"] != "PENDING":
            return result

        # 3. 轮询查询
        print("支付处理中，启动轮询查询...")
        start_time = time.time()

        while True:
            elapsed = time.time() - start_time

            # 超时退出
            if elapsed > max_wait_seconds:
                print(f"轮询超时（{max_wait_seconds}秒），请人工确认订单状态")
                return {"status": "TIMEOUT", "data": {"client_sn": client_sn}, "message": "轮询超时"}

            # 轮询间隔：前60秒3秒，之后10秒
            interval = 3 if elapsed < 60 else 10
            time.sleep(interval)

            # 查询订单
            query_result = self.query(client_sn=client_sn)
            if query_result.get("result_code") != "200":
                print(f"查询失败，继续重试...")
                continue

            order_status = query_result.get("biz_response", {}).get("data", {}).get("order_status", "")
            print(f"[{elapsed:.0f}s] 订单状态: {order_status}")

            # 最终状态判定
            if order_status == "PAID":
                data = query_result["biz_response"]["data"]
                return {"status": "SUCCESS", "data": data, "message": "支付成功"}
            elif order_status == "PAY_CANCELED":
                return {"status": "FAIL", "data": None, "message": "支付已撤销"}
            # 其他状态继续轮询


if __name__ == "__main__":
    client = ShouqianbaPayClient(
        terminal_sn="your_terminal_sn",
        terminal_key="your_terminal_key",
    )

    # 生成唯一订单号
    client_sn = generate_client_sn("STORE01")

    # 发起支付（带轮询）
    # 注意：这是真实交易！测试后请务必退款
    result = client.pay_with_polling(
        client_sn=client_sn,
        total_amount="1",                  # 1分钱（测试用）
        dynamic_id="130818341921600584",   # 付款码（扫码枪获取）
        subject="测试商品",                 # 账单显示
        operator="cashier_01",             # 操作员
    )

    print(f"\n支付结果: {result}")

File: sqb-api-skills/sqb-pay/reference/PayExample.java
package com.example.shouqianba;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.concurrent.TimeUnit;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

/**
 * 收钱吧 B扫C 付款码支付示例
 *
 * 核心流程：扫码 → 支付 → 轮询查询 → 确认结果
 * 注意：所有交易都是真实的（无沙盒环境），测试后务必退款
 */
public class PayExample {

    private static final String API_BASE = "https://vsi-api.shouqianba.com";
    private static final MediaType JSON_TYPE = MediaType.get("application/json; charset=utf-8");
    private static final OkHttpClient client = new OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build();
    private static final ObjectMapper mapper = new ObjectMapper();

    // 终端凭证（激活后获得，替换为真实值）
    private String terminalSn;
    private String terminalKey;

    public PayExample(String terminalSn, String terminalKey) {
        this.terminalSn = terminalSn;
        this.terminalKey = terminalKey;
    }

    /**
     * 付款码支付
     *
     * @param clientSn   商户唯一订单号
     * @param totalAmount 金额（单位：分）
     * @param dynamicId  顾客付款码内容（扫码枪获取）
     * @param subject    交易简介（显示在顾客账单）
     * @param operator   操作员
     * @return 支付结果
     */
    public PayResult pay(String clientSn, String totalAmount, String dynamicId,
                         String subject, String operator) throws IOException {

        // 1. 构建请求体
        ObjectNode body = mapper.createObjectNode();
        body.put("terminal_sn", terminalSn);
        body.put("client_sn", clientSn);
        body.put("total_amount", totalAmount);
        body.put("dynamic_id", dynamicId);
        body.put("subject", subject);
        body.put("operator", operator);

        String bodyStr = mapper.writeValueAsString(body);

        // 2. 计算签名
        String sign = md5(bodyStr + terminalKey);

        // 3. 发送支付请求
        Request request = new Request.Builder()
            .url(API_BASE + "/upay/v2/pay")
            .addHeader("Authorization", terminalSn + " " + sign)
            .addHeader("Content-Type", "application/json; charset=utf-8")
            .post(RequestBody.create(bodyStr, JSON_TYPE))
            .build();

        try (Response response = client.newCall(request).execute()) {
            String respBody = response.body().string();
            JsonNode resp = mapper.readTree(respBody);

            // 4. 三层状态判定
            return parsePayResponse(resp, clientSn);
        }
    }

    /**
     * 解析支付响应（三层判定）
     */
    private PayResult parsePayResponse(JsonNode resp, String clientSn) {
        // 第一层：通信层
        String resultCode = resp.path("result_code").asText();
        if (!"200".equals(resultCode)) {
            return new PayResult("FAIL", null,
                "通信失败: " + resp.path("error_message").asText());
        }

        JsonNode bizResponse = resp.path("biz_response");
        String bizResultCode = bizResponse.path("result_code").asText();

        // 第二层：业务层
        switch (bizResultCode) {
            case "PAY_SUCCESS":
                break; // 继续判断第三层
            case "PAY_FAIL":
                return new PayResult("FAIL", null, "支付失败");
            case "PAY_IN_PROGRESS":
            case "PAY_FAIL_ERROR":
                // 非最终状态，需要轮询
                return new PayResult("PENDING", clientSn, "支付处理中，需轮询查询");
            default:
                return new PayResult("PENDING", clientSn, "未知业务状态: " + bizResultCode);
        }

        // 第三层：订单状态
        JsonNode data = bizResponse.path("data");
        String orderStatus = data.path("order_status").asText();

        switch (orderStatus) {
            case "PAID":
                return new PayResult("SUCCESS", data.path("sn").asText(), "支付成功");
            case "PAY_CANCELED":
                return new PayResult("FAIL", null, "支付已撤销");
            default:
                // CREATED, PAY_ERROR 等非最终状态
                return new PayResult("PENDING", clientSn, "状态: " + orderStatus + "，需轮询");
        }
    }

    /**
     * 带轮询的完整支付流程
     */
    public PayResult payWithPolling(String clientSn, String totalAmount, String dynamicId,
                                     String subject, String operator) throws IOException, InterruptedException {
        // 1. 发起支付
        PayResult result = pay(clientSn, totalAmount, dynamicId, subject, operator);

        // 2. 如果是最终状态，直接返回
        if (!"PENDING".equals(result.status)) {
            return result;
        }

        // 3. 非最终状态，启动轮询
        System.out.println("支付处理中，启动轮询查询...");
        long startTime = System.currentTimeMillis();
        int maxWaitSeconds = 120;

        while (true) {
            long elapsed = (System.currentTimeMillis() - startTime) / 1000;

            // 超时退出
            if (elapsed > maxWaitSeconds) {
                System.err.println("轮询超时（" + maxWaitSeconds + "秒），请人工确认订单状态");
                return new PayResult("TIMEOUT", clientSn, "轮询超时");
            }

            // 轮询间隔：前60秒3秒，之后10秒
            int interval = elapsed < 60 ? 3 : 10;
            Thread.sleep(interval * 1000L);

            // 查询订单
            JsonNode queryResult = queryOrder(clientSn);
            if (queryResult == null) {
                continue; // 查询失败，继续重试
            }

            String orderStatus = queryResult.path("biz_response")
                .path("data").path("order_status").asText();

            System.out.println("[" + elapsed + "s] 订单状态: " + orderStatus);

            // 判断最终状态
            if ("PAID".equals(orderStatus)) {
                return new PayResult("SUCCESS",
                    queryResult.path("biz_response").path("data").path("sn").asText(),
                    "支付成功");
            } else if ("PAY_CANCELED".equals(orderStatus)) {
                return new PayResult("FAIL", null, "支付已撤销");
            }
            // 其他状态继续轮询
        }
    }

    /**
     * 查询订单（用于轮询）
     */
    private JsonNode queryOrder(String clientSn) throws IOException {
        ObjectNode body = mapper.createObjectNode();
        body.put("terminal_sn", terminalSn);
        body.put("client_sn", clientSn);

        String bodyStr = mapper.writeValueAsString(body);
        String sign = md5(bodyStr + terminalKey);

        Request request = new Request.Builder()
            .url(API_BASE + "/upay/v2/query")
            .addHeader("Authorization", terminalSn + " " + sign)
            .addHeader("Content-Type", "application/json; charset=utf-8")
            .post(RequestBody.create(bodyStr, JSON_TYPE))
            .build();

        try (Response response = client.newCall(request).execute()) {
            return mapper.readTree(response.body().string());
        }
    }

    /**
     * 生成唯一订单号
     */
    public static String generateClientSn(String storeId) {
        String timestamp = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMddHHmmssSSS"));
        return storeId + timestamp + String.format("%04d", (int) (Math.random() * 10000));
    }

    private static String md5(String input) {
        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            byte[] digest = md.digest(input.getBytes(StandardCharsets.UTF_8));
            StringBuilder sb = new StringBuilder();
            for (byte b : digest) {
                sb.append(String.format("%02x", b));
            }
            return sb.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * 支付结果
     */
    public static class PayResult {
        public final String status;   // SUCCESS, FAIL, PENDING, TIMEOUT
        public final String orderSn;  // 收钱吧订单号或 client_sn
        public final String message;

        public PayResult(String status, String orderSn, String message) {
            this.status = status;
            this.orderSn = orderSn;
            this.message = message;
        }

        @Override
        public String toString() {
            return "PayResult{status='" + status + "', orderSn='" + orderSn
                + "', message='" + message + "'}";
        }
    }

    public static void main(String[] args) throws Exception {
        PayExample payApi = new PayExample("your_terminal_sn", "your_terminal_key");

        // 生成唯一订单号
        String clientSn = generateClientSn("STORE01");

        // 发起支付（带轮询）
        // 注意：这是真实交易！测试后请务必退款
        PayResult result = payApi.payWithPolling(
            clientSn,
            "1",                     // 1分钱（测试用）
            "130818341921600584",     // 付款码（扫码枪获取）
            "测试商品",               // 账单显示
            "cashier_01"             // 操作员
        );

        System.out.println("支付结果: " + result);
    }
}
